### ISSUE TITLE
KEDA metric server is leaking tcp connections

### ISSUE THREAD

**Issue by nikhil-bhat** [kedacore/keda#3387]
### Report

we noticed following errors E0716 16:51:54.766808 1 provider.go:124] keda_metrics_adapter/provider "msg"="error getting metric for scaler" "error"="Get \"http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=redacted\": dial tcp 169.254.169.254:80: connect: cannot assign requested address".

after entering the network namespaces of the metrics adaptor we found keda-mteric0-adaptor had opened 28k connections which is the limit and for new connections we were getting the  error mentioned above.

We are using keda 2.7.1 in our cluster
![image](https://user-images.githubusercontent.com/6253611/179494138-59b9bdeb-392f-48aa-9179-ae88d3510218.png)

### Expected Behavior

keda should not leak tcp connections

### Actual Behavior

keda is creating new tcp connections

### Steps to Reproduce the Problem

1. Install keda 2.7.1
2. 
3.

### Logs from KEDA operator

```
E0716 16:51:54.766808 1 provider.go:124] keda_metrics_adapter/provider "msg"="error getting metric for scaler" "error"="Get \"http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=redacted\": dial tcp 169.254.169.254:80: connect: cannot assign requested address"
```

### KEDA Version
2.7.1

### Kubernetes Version

1.21.10

### Platform

azure

### Scaler Details

azure_storage azure service bus

**Comment by JorTurFer** [comment]
hey,
we should address it, are you willing to contribute? There are other issue related and maybe the problem in both cases comes from the same code https://github.com/kedacore/keda/issues/3151#issuecomment-1184443468

**Comment by JorTurFer** [comment]
I don't want to put pressure in your back, but just for giving the whole picture. We will do a release during the first half of August and having this fix would be nice.
As I said, don't feel it as a pressure, we have plans to check this before the release, so if you don't have time it's not problem at all

**Comment by zroubalik** [comment]
@nikhil-bhat would you mind trying a KEDA distribution with a proposed fix to confirm whether the issue has been fixed or not?

**Comment by JorTurFer** [comment]
that's true, maybe the problem in Azure scalers it's different and related with not closed connections 🤔

**Comment by nikhil-bhat** [comment]
maybe what we need is this http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#close_http_resp_body

**Comment by nikhil-bhat** [comment]
mostly imds @amirschw

**Comment by amirschw** [comment]
It's worth mentioning that we didn't notice this behavior in previous versions (we upgraded from 2.4 directly to 2.7.1). Another thing we noticed after the upgrade was a big increase in the rate of token requests sent to IMDS from both the metrics server and the operator.

**Comment by zroubalik** [comment]
> > so, in 2.4 you didn't have this problem, and now in 2.7.1 you have it, right?
> 
> right, another theory we have is that this is only happening when there are scalers that are getting 403

interesting, that would mean that this might be caused by the same problem as described here: https://github.com/kedacore/keda/issues/3151

**Comment by timown** [comment]
happens also with keda operator

**Comment by timown** [comment]
still an issue

**Comment by timown** [comment]
still an issue

**Comment by amirschw** [comment]
Can we keep this alive please? @zroubalik @JorTurFer

**Comment by radekfojtik** [comment]
Hi guys, we also encountered this problem and I can confirm the assumption that the cause is scaler that gets 403 (that was our case - kedacore/keda 2.8.1)

**Comment by radekfojtik** [comment]
> @radekfojtik could you please tell me which scaler are you using? And by chance, do you think you can try more up to date version?

In this case it was 'azure-queue' (keda-operator did not have access to the storage account). Unfortunately I don't have the capacity right now to test it on a newer version.

**Comment by RomanDvorsky** [comment]
Hi, 
we have just installed the latest version `2.10.1` and I have suspicion that the leak is still there.

here is the last error log (it was fixed that time):
![image](https://github.com/kedacore/keda/assets/36188474/d41c7272-f30d-4162-bdd5-b952270f33d8)

here is the consumed memory (there is 1 hours shift):
![image](https://github.com/kedacore/keda/assets/36188474/896a13e4-b1f5-42f0-b638-70f58c84ecc2)
the blue line in the graph is the memory request.

I will try to check later today whether we can perform one more test which will last bit longer.

Regards,
R.

**Comment by zroubalik** [comment]
Thanks for the details folks.
